Microsoft Agents Framework mini workshop (#551)
neo4j-graphacademy/courses: asciidoc/courses/workshop-mini-maf/modules/1-introduction/lessons/1-what-is-the-framework/lesson.adoc

Reviewer: Why not just have a single module with 3 lessons? Adding modules introduces unnecessary clicks and boiler plate for little benefit.

= What is Microsoft Agent Framework?
:type: lesson
:order: 1
:slides: true

In this lesson, you will learn:

* What Microsoft Agent Framework is and when teams adopt it
* How MAF relates to Semantic Kernel and AutoGen
* How MAF-style development fits next to the GenAI workshop

[.slide.discrete]
== Introduction

> **Microsoft Agent Framework** is an open, cross-language SDK for building **AI agents** and **multi-agent** systems on top of modern language models. Microsoft positions it as a single place to model agents, connect **tools**, run **orchestrated workflows**, and integrate with protocols such as **Model Context Protocol (MCP)**—the same protocol many Neo4j and GraphAcademy labs use for tool calling.

Reviewer: ```suggestion
**Microsoft Agent Framework (MAF)** is an open, cross-language SDK for building **AI agents** and **multi-agent** systems on top of modern language models. Microsoft positions it as a single place to model agents, connect **tools**, run **orchestrated workflows**, and integrate with protocols such as **Model Context Protocol (MCP)**—the same protocol many Neo4j and GraphAcademy labs use for tool calling.
```

> modern language models

such as? 

> the same protocol many Neo4j and GraphAcademy labs use for tool calling.

do they? such as?

> This lesson stays conceptual. After link:/courses/workshop-genai/[Neo4j and Generative AI Workshop^], you already know retrieval, embeddings, and conversational patterns. Here you add a **runtime and library lens**: how a mainstream agent framework packages those ideas for production-style applications.

Reviewer: The language feels disjointed.

```suggestion
This lesson builds on your existing knowledge of retrieval, embeddings, and conversational patterns. You will explore how a mainstream agent framework packages those ideas for production-style applications.

[TIP]
.Neo4j & Generative AI Fundamentals
====
You can learn about Generative AI, retrievers, embeddings, vectors and more in the link:/courses/workshop-genai[Neo4j & Generative AI GraphAcademy course^]
====
```

> After link:/courses/workshop-genai/[Neo4j and Generative AI Workshop^],

Is the expectation that they will have completed the GenAI workshop? If we are referencing pre-requisite courses, we should direct them to the Neo4j & Generative AI Fundamentals online course.

Author: This workshop is meant to be delivered as an additional module for the GenAI workshop

[.slide]
== Why a dedicated framework?

> Teams outgrow one-off scripts when they need **consistent** patterns for:

Reviewer: what do we mean by scripts? programs?

Author: I don't see the need of overexplaining what a script is

* **Tool use** (calling APIs, databases, or MCP servers with structured inputs)
* **Multi-step workflows** (branching, retries, human approval)
* **Multiple agents** (specialists that hand off work or run in parallel)
* **Observability** (streaming, logging, checkpoints)

A framework does not replace good application design. It gives you **shared vocabulary**, **reusable components**, and **documentation** so more than one developer can maintain the system.

[.slide]
== Lineage and scope

> Microsoft Agent Framework builds on ideas from **Semantic Kernel** and **AutoGen**, combining orchestration and agent abstractions into a **unified** programming model for **Python** and **.NET**. The exact feature set evolves with each release; treat the official repository and release notes as the source of truth for API names and package layout.

Reviewer: ```suggestion
Microsoft Agent Framework builds on ideas from **Semantic Kernel** and **AutoGen**, combining orchestration and agent abstractions into a **unified** programming model for **Python** and **.NET**.

[IMPORTANT]
.New features
====
The exact feature set evolves with each release; treat the official repository and release notes as the source of truth for API names and package layout.
====
```

> Expect **multi-provider** model support (commercial and open weights), **graph-structured workflows** in higher-level samples, and emphasis on **interoperability** with MCP and agent-to-agent patterns—both relevant when Neo4j appears as a tool or knowledge source.

Reviewer: This could be split into 2 sentences to make it easier to read.

[.slide]
== How you work with MAF in practice

> Typical steps (at a high level):

Reviewer: ```suggestion
High-level steps:
```

> . Pick a **language runtime** (Python or .NET) and add the published packages from the official instructions.

Reviewer: ```suggestion
. Pick a **language runtime** (Python or .NET) and add the published packages.
```

. Define one or more **agents** with instructions and allowed **tools**.
. Run **single-turn** or **multi-turn** flows; scale up to **orchestrated** graphs when the task splits cleanly across steps or roles.

You still choose **where truth lives**—documents, a graph database, vector indexes—and how retrieval runs. MAF helps you **wire** models, tools, and control flow in code.

[.next]
== Next

read::Continue[]

[.summary]
== Summary

> In this lesson, you learned what Microsoft Agent Framework is, why teams adopt a framework instead of only scripts, and how it relates to Semantic Kernel and AutoGen. You also outlined how MAF-style development fits next to the GenAI workshop’s retrieval and agent concepts.

Reviewer: >and how it relates to Semantic Kernel and AutoGen

Did I learn this? It is mentioned but I dont think I learnt this.

>You also outlined how MAF-style development fits next to the GenAI workshop’s retrieval and agent concepts.

I did outline anything. Did I learn this?
